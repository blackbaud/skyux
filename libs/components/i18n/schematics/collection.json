{
  "$schema": "../../../../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "ng-add": {
      "description": "Add @skyux/i18n to your project",
      "factory": "./ng-add/ng-add.schematic"
    },
    "lib-resources-module": {
      "description": "Generate library resources module",
      "factory": "./ng-generate/lib-resources-module/lib-resources-module.schematic",
      "schema": "./ng-generate/lib-resources-module/schema.json"
    }
  }
}
