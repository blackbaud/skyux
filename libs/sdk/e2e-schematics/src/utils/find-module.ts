import { Tree, normalizePath, visitNotIgnoredFiles } from '@nx/devkit';
import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

import { relative } from 'path';

import {
  DecoratedClass,
  findComponentClass,
  findImport,
  findNgModuleClass,
  getNamedImport,
  readSourceFile,
} from './ast-utils';
import { dirname } from './utils';

export function findModulePaths(
  tree: Tree,
  path: string,
  predicate: (path: string) => boolean,
  endsWith = '.module.ts'
): string[] {
  const modulePaths: string[] = [];
  visitNotIgnoredFiles(tree, path, (filePath) => {
    if (filePath.endsWith(endsWith)) {
      if (predicate(filePath)) {
        modulePaths.push(normalizePath(filePath));
      }
    }
  });
  return modulePaths;
}

export function findDeclaringModule(
  tree: Tree,
  path: string,
  componentPath: string
): { filepath: string; module: DecoratedClass } | null {
  let result: { filepath: string; module: DecoratedClass } | null = null;
  const componentSource = readSourceFile(tree, componentPath);
  const componentClass = findComponentClass(componentSource);
  if (componentClass) {
    findModulePaths(
      tree,
      path,
      (filepath) => {
        filepath = normalizePath(filepath);
        const filepathDirectory = filepath.substring(
          0,
          filepath.lastIndexOf('/')
        );
        let relativePath = normalizePath(
          relative(filepathDirectory, componentPath)
        );
        if (!relativePath.startsWith('.')) {
          relativePath = `./${relativePath}`;
        }
        try {
          const sourceFile = readSourceFile(tree, filepath);
          const componentFileImport = findImport(sourceFile, relativePath);
          if (!componentFileImport) {
            return false;
          }
          const componentNamedImport = getNamedImport(
            componentFileImport,
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            componentClass.classDeclaration!.name!.text
          );
          const module = findNgModuleClass(sourceFile);
          if (module) {
            if (
              module.properties['declarations'] &&
              ts.isArrayLiteralExpression(module.properties['declarations'])
            ) {
              const declarations = module.properties['declarations'].elements;
              for (const declaration of declarations) {
                if (
                  ts.isIdentifier(declaration) &&
                  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                  declaration.text === componentNamedImport!.name.text
                ) {
                  result = { filepath, module };
                  return true;
                }
              }
            }
          }
          // eslint-disable-next-line no-empty
        } catch (e) {}
        return false;
      },
      '.module.ts'
    );
  }
  return result;
}

export function isRoutingModule(
  module: DecoratedClass,
  sourceFile: ts.SourceFile
): boolean {
  // Is one of the imports a static call to a method on RouterModule?
  if (
    module.properties['imports'] &&
    ts.isArrayLiteralExpression(module.properties['imports'])
  ) {
    if (
      (module.properties['imports'] as ts.ArrayLiteralExpression).elements.some(
        (im) => {
          if (
            ts.isCallExpression(im) &&
            ts.isPropertyAccessExpression(im.getChildren(sourceFile)[0])
          ) {
            const propertyAccessExpression = im.getChildren(
              sourceFile
            )[0] as ts.PropertyAccessExpression;
            if (
              ts.isIdentifier(
                (propertyAccessExpression as ts.PropertyAccessExpression)
                  .expression
              )
            ) {
              return (
                (propertyAccessExpression.expression as ts.Identifier).text ===
                'RouterModule'
              );
            }
          }
          return false;
        }
      )
    ) {
      return true;
    }
  }
  return false;
}

export function findClosestModule(
  modulePaths: string[],
  projectDirectory: string,
  closestToDirectory: string
): string | undefined {
  projectDirectory = normalizePath(projectDirectory);
  closestToDirectory = normalizePath(closestToDirectory);
  const modulePathsNormalized = modulePaths.map((modulePath) =>
    normalizePath(modulePath)
  );
  modulePathsNormalized.sort((a, b) => {
    const aDir = dirname(a);
    const bDir = dirname(b);
    if (aDir.length === bDir.length) {
      return a > b ? 1 : -1;
    }
    return aDir.length < bDir.length ? 1 : -1;
  });
  let directory = `${projectDirectory}/${closestToDirectory}`;
  while (
    directory &&
    directory !== projectDirectory &&
    !modulePathsNormalized.some((m) => dirname(m) === directory)
  ) {
    directory = dirname(directory);
  }
  if (!directory || directory === projectDirectory) {
    return undefined;
  }
  // Return the basename of the module.
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return modulePathsNormalized
    .find((m) => dirname(m) === directory)!
    .split('/')
    .pop()!
    .replace(/\.module\.ts$/, '');
}
