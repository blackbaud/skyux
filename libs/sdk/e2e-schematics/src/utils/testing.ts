import {
  E2eTestRunner,
  UnitTestRunner,
  applicationGenerator,
  libraryGenerator,
} from '@nx/angular/generators';
import { Tree } from '@nx/devkit';
import { Linter } from '@nx/eslint';

export async function createTestApplication(
  tree: Tree,
  options: { name: string; e2eTestRunner?: boolean; unitTestRunner?: boolean },
): Promise<void> {
  await applicationGenerator(tree, {
    name: options.name,
    skipPackageJson: true,
    directory: `apps/${options.name}`,
    e2eTestRunner: options.e2eTestRunner
      ? E2eTestRunner.Cypress
      : E2eTestRunner.None,
    linter: Linter.None,
    unitTestRunner: options.unitTestRunner
      ? UnitTestRunner.Jest
      : UnitTestRunner.None,
  });
}

export async function createTestLibrary(
  tree: Tree,
  options: { name: string },
): Promise<void> {
  await libraryGenerator(tree, {
    name: options.name,
    skipPackageJson: true,
    directory: `libs/${options.name}`,
    linter: Linter.None,
    unitTestRunner: UnitTestRunner.None,
  });
}
