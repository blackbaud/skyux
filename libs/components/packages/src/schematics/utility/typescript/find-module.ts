import { Tree } from '@angular-devkit/schematics';
import {
  getDecoratorMetadata,
  getMetadataField,
  isImported,
  parseSourceFile,
} from '@angular/cdk/schematics';
import ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';
import { buildRelativePath } from '@schematics/angular/utility/find-module';

import { visitProjectFiles } from '../visit-project-files';

function findModuleInPath(
  tree: Tree,
  path: string,
  filter: (sourceFile: ts.SourceFile) => boolean,
): {
  filepath: string;
  module: { className: string; metadata: ts.ObjectLiteralElement };
} | null {
  let found: {
    filepath: string;
    module: { className: string; metadata: ts.ObjectLiteralElement };
  } | null = null;
  visitProjectFiles(tree, path, (filePath) => {
    if (found) {
      return; // Stop searching if we already found a module
    }
    if (filePath.endsWith('.ts')) {
      const sourceFile = parseSourceFile(tree, filePath);
      if (
        isImported(sourceFile, 'NgModule', '@angular/core') &&
        filter(sourceFile)
      ) {
        const metadata = getDecoratorMetadata(
          sourceFile,
          'NgModule',
          '@angular/core',
        )[0] as ts.ObjectLiteralElement;
        const className =
          findClassDeclarationParent(metadata)?.name?.text || '';
        found = {
          filepath: buildRelativePath('/', `/${filePath}`),
          module: {
            className,
            metadata,
          },
        };
      }
    }
  });
  return found;
}

function findClassDeclarationParent(
  node: ts.Node | undefined,
): ts.ClassDeclaration | undefined {
  if (node && ts.isClassDeclaration(node)) {
    return node;
  }
  if (node?.parent) {
    return findClassDeclarationParent(node.parent);
  }
  return undefined;
}

export function isStandaloneComponent(
  metadata: ts.ObjectLiteralExpression,
): boolean {
  const standalone = getMetadataField(metadata, 'standalone');
  return (
    !standalone[0] ||
    (ts.isPropertyAssignment(standalone[0]) &&
      standalone[0].initializer.kind === ts.SyntaxKind.TrueKeyword)
  );
}

export function findDeclaringModule(
  tree: Tree,
  path: string,
  componentPath: string,
): {
  filepath: string;
  module: { className: string; metadata: ts.ObjectLiteralElement };
} | null {
  const source = parseSourceFile(tree, componentPath);
  const metadata = getDecoratorMetadata(
    source,
    'Component',
    '@angular/core',
  )[0] as ts.ObjectLiteralElement | undefined;
  const componentClass = findClassDeclarationParent(metadata);

  const componentClassName = componentClass?.name?.text;
  if (
    componentClassName &&
    metadata &&
    ts.isObjectLiteralExpression(metadata)
  ) {
    if (isStandaloneComponent(metadata)) {
      return {
        filepath: buildRelativePath('/', `/${componentPath}`),
        module: {
          className: componentClass.name.text,
          metadata: metadata,
        },
      };
    }

    return findModuleInPath(tree, path, (sourceFile) => {
      const moduleMetadata = getDecoratorMetadata(
        sourceFile,
        'NgModule',
        '@angular/core',
      )[0] as ts.ObjectLiteralExpression | undefined;
      if (!moduleMetadata) {
        return false;
      }
      const declarations = getMetadataField(moduleMetadata, 'declarations');
      return (
        declarations &&
        declarations.some(
          (declaration) =>
            ts.isPropertyAssignment(declaration) &&
            ts.isArrayLiteralExpression(declaration.initializer) &&
            declaration.initializer.elements.some(
              (item) => item.getText(sourceFile) === componentClassName,
            ),
        )
      );
    });
  }

  return null;
}
