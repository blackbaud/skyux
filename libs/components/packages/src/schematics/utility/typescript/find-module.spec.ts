import { SchematicTestRunner } from '@angular-devkit/schematics/testing';

import {
  angularComponentGenerator,
  angularModuleGenerator,
} from '../../testing/angular-module-generator';
import { createTestApp } from '../../testing/scaffold';

import { findDeclaringModule, isStandaloneComponent } from './find-module';

describe('find-module', () => {
  const runner = new SchematicTestRunner(
    'schematics',
    require.resolve('../../../../collection.json'),
  );

  it('should find module', async () => {
    const tree = await createTestApp(runner, {
      projectName: 'test',
    });
    await angularModuleGenerator(runner, tree, {
      name: 'test',
      project: 'test',
    });
    await angularModuleGenerator(runner, tree, {
      name: 'other',
      project: 'test',
    });
    await angularComponentGenerator(runner, tree, {
      name: 'test',
      project: 'test',
      module: 'test',
      standalone: false,
    });
    expect(tree.readText('src/app/test/test.module.ts')).toMatchSnapshot();
    const module = findDeclaringModule(
      tree,
      'src',
      'src/app/test/test.component.ts',
    );
    expect(module?.filepath).toBe('./src/app/test/test.module.ts');
    expect(module?.module.className).toBe('TestModule');
  });

  it('should find standalone component', async () => {
    const tree = await createTestApp(runner, {
      projectName: 'test',
    });
    await angularModuleGenerator(runner, tree, {
      name: 'test',
      project: 'test',
    });
    await angularComponentGenerator(runner, tree, {
      name: 'test',
      project: 'test',
      module: 'test',
    });
    expect(tree.readText('src/app/test/test.module.ts')).toMatchSnapshot();
    const module = findDeclaringModule(
      tree,
      'src',
      'src/app/test/test.component.ts',
    );
    expect(module?.filepath).toBe('./src/app/test/test.component.ts');
    expect(module?.module.className).toBe('TestComponent');
    expect(module?.module.metadata).toBeTruthy();
    expect(isStandaloneComponent(module?.module.metadata as any)).toBe(true);
  });

  it('should find module, skip non-declaring reference', async () => {
    const tree = await createTestApp(runner, {
      projectName: 'test',
    });
    await angularModuleGenerator(runner, tree, {
      name: 'test/bogus',
      flat: true,
      project: 'test',
    });
    const recorder = tree.beginUpdate('src/app/test/bogus.module.ts');
    recorder.insertLeft(
      0,
      `import { TestComponent } from './test.component';\n`,
    );
    await angularModuleGenerator(runner, tree, {
      name: 'test',
      project: 'test',
    });
    await angularComponentGenerator(runner, tree, {
      name: 'test',
      project: 'test',
      module: 'test',
      standalone: false,
    });
    const module = findDeclaringModule(
      tree,
      'src',
      'src/app/test/test.component.ts',
    );
    expect(module?.filepath).toBe('./src/app/test/test.module.ts');
    expect(module?.module.className).toBe('TestModule');
  });
});
