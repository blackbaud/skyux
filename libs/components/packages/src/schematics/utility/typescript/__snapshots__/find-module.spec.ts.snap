// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-module should find module 1`] = `
"import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TestComponent } from './test.component';



@NgModule({
  declarations: [
    TestComponent
  ],
  imports: [
    CommonModule
  ]
})
export class TestModule { }
"
`;

exports[`find-module should find standalone component 1`] = `
"import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';



@NgModule({
  declarations: [],
  imports: [
    CommonModule
  ]
})
export class TestModule { }
"
`;
